### PR TITLE
ZStream interface for the consumer

### LINKED ISSUE #8: Add a ZIO Streams-based interface for interacting with the consumer
For starters, we need an interface that will allow consuming the consumer's data as a `StreamChunk[Throwable, Record[K, V]]`, where `Record` is something like:
```scala
case class Record[K, V](key: K, value: V, timestamp: Long, metadata: Metadata)
case class Metadata(topicPartition: TopicPartition, offset: Long)
```

An important part for achieving good throughput is decoupling the consumption from processing:
- The consumer must be polled in a separate fiber from the rest of the stream with the data transferred to the stream via a bounded queue
- When the queue is full, the partitions on the consumer must be paused and polling must continue (or the consumer will be removed from the consumer group)
- Commits should be fed back to the consumer fiber via another queue and be processed by the polling fiber

See https://github.com/iravid/fs2-kafka-streams/blob/master/src/main/scala/com/iravid/fs2/kafka/client/RecordStream.scala#L262 for a previous implementation.

### PR DESCRIPTION
**Pull request by iravid**
Resolves #8.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -24,8 +24,13 @@ developers := List(
   )
 )
 
+testFrameworks += new TestFramework("zio.test.sbt.ZTestFramework")
+
 libraryDependencies ++= Seq(
-  "dev.zio"                 %% "zio"            % "1.0.0-RC10-1",
+  "dev.zio"                 %% "zio"            % "1.0.0-RC12-1",
```
[the remark this answers is not in the document]
Comment (iravid) [reply build.sbt]: Good point.

File: src/main/scala/zio/kafka/client/Runloop.scala [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+                  }.catchAll(e => cont(Exit.fail(e)))
+                }
+          } yield state
+
+        cmd match {
+          case Command.Poll() =>
+            val pendingRequests = state.pendingRequests
+            for {
+              pollResult <- deps.consumer.withConsumerM { c =>
+                             ZIO {
```
[the remark this answers is not in the document]
Comment (iravid) [reply src/main/scala/zio/kafka/client/Runloop.scala]: I defer to the scalafmt overlords

File: src/main/scala/zio/kafka/client/ConsumerSettings.scala [hunk 3]
```diff
@@ -8,13 +8,17 @@ case class ConsumerSettings(
   groupId: String,
   clientId: String,
   closeTimeout: Duration,
-  extraDriverSettings: Map[String, String]
+  extraDriverSettings: Map[String, String],
```
[the remark this answers is not in the document]
Comment (iravid) [reply src/main/scala/zio/kafka/client/ConsumerSettings.scala]: Yes, good point!

File: src/main/scala/zio/kafka/client/CommittableRecord.scala [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package zio.kafka.client
+
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.common.TopicPartition
+import zio.Task
+
+case class CommittableRecord[K, V](record: ConsumerRecord[K, V], offset: Offset)
+object CommittableRecord {
+  def fromConsumerRecord[K, V](
```
[the remark this answers is not in the document]
Comment (iravid) [reply src/main/scala/zio/kafka/client/CommittableRecord.scala]: Yup

**Comment by iravid** [comment]
@NeQuissimus Good points there -- I need to test how resilient this setup to failures in the consumer. I'll do this in follow-ups.

File: src/main/scala/zio/kafka/client/Runloop.scala [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+  sealed abstract class Rebalance
+  object Rebalance {
+    case class Revoked(previousAssignment: Set[TopicPartition]) extends Rebalance
+    case class Assigned(currentAssignment: Set[TopicPartition]) extends Rebalance
+  }
+
+  case class Deps[K, V](
+    consumer: ConsumerAccess[K, V],
+    pollFrequency: Duration,
```
Comment (pierangeloc) [inline src/main/scala/zio/kafka/client/Runloop.scala]: `pollInterval` is probably a more accurate name than `pollFrequency`

File: src/main/scala/zio/kafka/client/Offset.scala [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  def commit: Task[Unit]
+  def batch: OffsetBatch
+}
+
+final case class OffsetImpl(
+  topicPartition: TopicPartition,
+  offset: Long,
+  commitHandle: Map[TopicPartition, Long] => Task[Unit]
+) extends Offset {
+  val commit: Task[Unit] = commitHandle(Map(topicPartition    -> (offset + 1)))
```
[the remark this answers is not in the document]
Comment (iravid) [reply src/main/scala/zio/kafka/client/Offset.scala]: When committing to Kafka, we need to commit the "next offset to be read" -- which is the next offset after the current offset.

I actually could bump the offset in the runloop - should read nicer and be less error prone.

File: src/main/scala/zio/kafka/client/OffsetBatch.scala [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    )
+
+  def merge(otherOffsets: OffsetBatch) = {
+    val newOffsets = Map.newBuilder[TopicPartition, Long]
+    newOffsets ++= offsets
+    otherOffsets.offsets.foreach {
+      case (tp, offset) =>
+        val existing = offsets.getOrElse(tp, -1L)
+        if (existing < offset)
+          newOffsets += tp -> offset
```
[the remark this answers is not in the document]
Comment (iravid) [reply src/main/scala/zio/kafka/client/OffsetBatch.scala]: No, the offset here has already been incremented. Good evidence that this logic should be moved into the runloop.

File: src/main/scala/zio/kafka/client/Runloop.scala (hunk 5 above)
Comment (iravid) [reply src/main/scala/zio/kafka/client/Runloop.scala]: 👍

File: src/main/scala/zio/kafka/client/Offset.scala [hunk 8]
```diff
@@ -0,0 +1,20 @@
+package zio.kafka.client
+
+import org.apache.kafka.common.TopicPartition
+import zio.Task
+
+sealed trait Offset {
```
[the remark this answers is not in the document]
Comment (iravid) [reply src/main/scala/zio/kafka/client/Offset.scala]: Nope

**Comment by iravid** [comment]
Merging this now - we can address subsequent comments in follow-ups

**Pull request merged**